coderd/workspaces_test.go

```diff
@@ -426,47 +426,441 @@ func TestWorkspace(t *testing.T) {
+				presets: []*proto.Preset{
+					{
+						Name: "test1",
+						Parameters: []*proto.PresetParameter{
+							{Name: "param1", Value: "value1"},
+						},
+					},
+					{
+						Name: "test2",
+						Parameters: []*proto.PresetParameter{
+							{Name: "param2", Value: "value2"},
+						},
+					},
+				},
+				templateVersionParameters: []*proto.RichParameter{
+					{Name: "param1", Type: "string", Required: false},
+					{Name: "param2", Type: "string", Required: false},
+				},
+				selectedPresetIndex: -1,
+			},
+		}
 
-		ctx := testutil.Context(t, testutil.WaitLong)
+		for _, tc := range testCases {
+			tc := tc // Capture range variable
+			t.Run(tc.name, func(t *testing.T) {
+				t.Parallel()
 
-		presets, err := client.TemplateVersionPresets(ctx, version.ID)
-		require.NoError(t, err)
-		require.Equal(t, 1, len(presets))
-		require.Equal(t, "test", presets[0].Name)
+				client, _, api := coderdtest.NewWithAPI(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
+				user := coderdtest.CreateFirstUser(t, client)
+				authz := coderdtest.AssertRBAC(t, api, client)
 
-		workspace := coderdtest.CreateWorkspace(t, client, template.ID, func(request *codersdk.CreateWorkspaceRequest) {
-			request.TemplateVersionPresetID = presets[0].ID
-		})
+				// Create a plan response with the specified presets and parameters
+				planResponse := &proto.Response{
+					Type: &proto.Response_Plan{
+						Plan: &proto.PlanComplete{
+							Presets:    tc.presets,
+							Parameters: tc.templateVersionParameters,
+						},
+					},
+				}
 
-		authz.Reset() // Reset all previous checks done in setup.
-		ws, err := client.Workspace(ctx, workspace.ID)
-		authz.AssertChecked(t, policy.ActionRead, ws)
-		require.NoError(t, err)
-		require.Equal(t, user.UserID, ws.LatestBuild.InitiatorID)
-		require.Equal(t, codersdk.BuildReasonInitiator, ws.LatestBuild.Reason)
-		require.Equal(t, presets[0].ID, *ws.LatestBuild.TemplateVersionPresetID)
+				version := coderdtest.CreateTemplateVersion(t, client, user.OrganizationID, &echo.Responses{
+					Parse:          echo.ParseComplete,
+					ProvisionPlan:  []*proto.Response{planResponse},
+					ProvisionApply: echo.ApplyComplete,
+				})
+				coderdtest.AwaitTemplateVersionJobCompleted(t, client, version.ID)
+				template := coderdtest.CreateTemplate(t, client, user.OrganizationID, version.ID)
 
-		org, err := client.Organization(ctx, ws.OrganizationID)
-		require.NoError(t, err)
-		require.Equal(t, ws.OrganizationName, org.Name)
+				ctx := testutil.Context(t, testutil.WaitLong)
+
+				// Check presets
+				presets, err := client.TemplateVersionPresets(ctx, version.ID)
+				require.NoError(t, err)
+				require.Equal(t, len(tc.presets), len(presets))
+
+				if len(tc.presets) > 0 {
+					// Verify preset names and parameters
+					for i, preset := range presets {
+						require.Equal(t, tc.presets[i].Name, preset.Name)
+
+						// Check if the preset should have parameters
+						if tc.presets[i].Parameters != nil {
```

Comment:
https://maelvls.dev/go-happy-line-of-sight/

tl;dr keep happy path unindented.

Reply:
I also think this check is unnecessary? This appears to be a slice of `*PresetParameter`, and that's safe to range over without the check.

https://go.dev/play/p/1rT5DI94VfK

Reply:
done.